### PR TITLE
bind when we need to only

### LINKED ISSUE #2389: Bind lazier
E.g. audit these
https://github.com/ampproject/amphtml/blob/bf39315ff22225c3e87c8224bfbfb260e08071e4/src/input.js#L42

Can we only bind when we actually listen or do we always listen?

### PR DESCRIPTION
**Pull request by chenshay**
close #2389

### CODE REVIEW

File: src/input.js [hunk 1]
```diff
@@ -84,7 +75,8 @@ export class Input {
     // mouse events.
     if (this.hasTouch_) {
       this.hasMouse_ = !this.hasTouch_;
-      listenOnce(win.document, 'mousemove', this.boundOnMouseMove_);
+      listenOnce(win.document, 'mousemove',
+        /** @const {!Function} */ (this.onMouseMove_.bind(this)));
```
Comment (erwinmombay) [inline src/input.js]: why do we need this cast and the const declaration?

**Comment by jridgewell** [comment]
These changes seem like a step backwards. Now, we'll bind each multiple times.

**Comment by erwinmombay** [comment]
agree with @jridgewell. maybe late bound + cache

File: src/input.js [hunk 2]
```diff
@@ -217,8 +235,11 @@ export class Input {
     // Repeat, if attempts allow.
     this.mouseConfirmAttemptCount_++;
     if (this.mouseConfirmAttemptCount_ <= MAX_MOUSE_CONFIRM_ATTEMPS_) {
-      listenOnce(this.win.document, 'mousemove',
-        /** @const {!Function} */ (this.onMouseMove_.bind(this)));
+      if (!this.boundOnMouseMove_) {
+        this.boundOnMouseMove_ =
```
Comment (jridgewell) [inline src/input.js]: This if will never be true.

File: src/input.js [hunk 3]
```diff
@@ -198,11 +210,17 @@ export class Input {
       this.mouseCanceled_();
       return undefined;
     }
+    if (!this.boundMouseCanceled_) {
+      this.boundMouseCanceled_ = this.mouseCanceled_.bind(this);
+    }
+    if (!this.boundMouseConfirmed_) {
```
Comment (jridgewell) [inline src/input.js]: Shouldn't need to check both.

File: src/input.js [hunk 4]
```diff
@@ -75,8 +84,11 @@ export class Input {
     // mouse events.
     if (this.hasTouch_) {
       this.hasMouse_ = !this.hasTouch_;
-      listenOnce(win.document, 'mousemove',
-        /** @const {!Function} */ (this.onMouseMove_.bind(this)));
+      if (!this.boundOnMouseMove_) {
```
Comment (jridgewell) [inline src/input.js]: No need to check.

File: src/input.js [hunk 5]
```diff
@@ -84,10 +84,8 @@ export class Input {
     // mouse events.
     if (this.hasTouch_) {
       this.hasMouse_ = !this.hasTouch_;
-      if (!this.boundOnMouseMove_) {
-        this.boundOnMouseMove_ =
-          /** @private {!Function} */ (this.onMouseMove_.bind(this));
-      }
+      this.boundOnMouseMove_ =
+        /** @private {!Function} */ (this.onMouseMove_.bind(this));
```
Comment (jridgewell) [inline src/input.js]: Curious why we need the type annotation.

Comment (erwinmombay) [inline src/input.js]: don't think we need it

Comment (chenshay) [inline src/input.js]: check-types fails if I don't have it:
found   : function (Event): (Promise|undefined)
required: (function ((Event|null)): ?|null)

**Pull request merged**